glommio/src/io/byte_slice_ext.rs

```diff
@@ -0,0 +1,120 @@
+    /// or `destination` doesn't have more space to hold them.
+    ///
+    /// ## Returns
+    ///
+    /// The amount of bytes written.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use glommio::LocalExecutor;
+    /// use glommio::io::DmaFile;
+    /// use glommio::io::ByteSliceExt;
+    ///
+    /// let ex = LocalExecutor::default();
+    /// ex.run(async {
+    ///     let file = DmaFile::create("test.txt").await.unwrap();
+    ///     let buf = file.alloc_dma_buffer(4096);
+    ///     let mut vec = vec![0; 64];
+    ///     let n = buf.read_at(0, &mut vec);
+    ///     assert_eq!(n, 64); // read 64 bytes, the size of the buffer
+    ///
+    ///     let n = buf.read_at(4090, &mut vec);
+    ///     assert_eq!(n, 6);  // read 6 bytes, as there are only 6 bytes left from this offset
+    ///     file.close().await.unwrap();
+    /// });
+    /// ```
+    fn read_at<D: AsMut<[u8]> + ?Sized>(&self, offset: usize, destination: &mut D) -> usize;
+}
+
+/// Utility methods for working with mutable byte slices/buffers.
+pub trait ByteSliceMutExt {
+    /// Writes data to this buffer from a user-provided byte slice, starting at a particular
+    /// offset
+    ///
+    /// ## Returns
+    ///
+    /// The amount of bytes written.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use glommio::LocalExecutor;
+    /// use glommio::io::DmaFile;
+    /// use glommio::io::ByteSliceMutExt;
+    ///
+    /// let ex = LocalExecutor::default();
+    /// ex.run(async {
+    ///     let file = DmaFile::create("test.txt").await.unwrap();
+    ///     let mut buf = file.alloc_dma_buffer(4096);
+    ///     let vec = vec![1; 64];
+    ///     let n = buf.write_at(0, &vec);
+    ///     assert_eq!(n, 64); // wrote 64 bytes.
+    ///
+    ///     let n = buf.write_at(4090, &vec);
+    ///     assert_eq!(n, 6);  // wrote 6 bytes, as there are only 6 bytes left from this offset
+    ///     file.close().await.unwrap();
+    /// });
+    /// ```
+    fn write_at<S: AsRef<[u8]> + ?Sized>(&mut self, offset: usize, src: &S) -> usize;
+
+    /// Writes the specified value into all bytes of this buffer
+    fn memset(&mut self, value: u8);
+}
+
+fn slice_read_at(source: &[u8], offset: usize, destination: &mut [u8]) -> usize {
+    if offset > source.len() {
+        0
+    } else {
+        let len = core::cmp::min(destination.len(), source.len() - offset);
+        destination[0..len].copy_from_slice(&source[offset..][..len]);
+        len
+    }
+}
+
+fn slice_write_at(destination: &mut [u8], offset: usize, source: &[u8]) -> usize {
+    if offset > destination.len() {
+        0
+    } else {
+        let len = std::cmp::min(source.len(), destination.len() - offset);
+        destination[offset..offset + len].copy_from_slice(&source[0..len]);
+        len
+    }
+}
+
+fn slice_memset(destination: &mut [u8], value: u8) {
+    unsafe { std::ptr::write_bytes(destination.as_mut_ptr(), value, destination.len()) }
```

Comment:
This can be replaced with a safe `for` loop, which shouldn't cause any regressions (rustc is probably good enough to optimize that to `memset` or inline it), but reduce the amount of `unsafe` code.

Reply:
should be easy to verify if this is indeed the case with `llvm-objdump`.

Reply:
That may be the case, but I have no experience in using `llvm-objdump`, so I'm unable to verify this.

Reply:
there's a way to cheat that which I find very handy. You can use rust playground, select release mode and instead of "build", you change the big red button to "show assembly"

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=3d184f0f5c56d5bf1b7c722728f41623

The unsafe code seems slightly more efficient:

```
playground::slice_memset:
	movq	%rsi, %rax
	movl	%edx, %esi
	movq	%rax, %rdx
	jmpq	*memset@GOTPCREL(%rip)
```

The code using for does generate a memset - which is good news - but it also wraps it in some boundary tests:
```
playground::slice_memset_for:
	testq	%rsi, %rsi
	je	.LBB1_2
	pushq	%rax
	movq	%rsi, %rax
	movzbl	%dl, %esi
	movq	%rax, %rdx
	callq	*memset@GOTPCREL(%rip)
	addq	$8, %rsp

.LBB1_2:
	retq
```

My reading is that it is just special casing the zero-size array first, and then the rest is about the same with one extra stack push.

It's a price worth paying to remove unsafe. memset is unlikely to be a super hot path and if it is a user can roll out theirn own

Reply:
ok, that also mean that currently, calling:
```rust
use glommio::io::ByteSliceMutExt;
[].memset(0);
```
introduces UB. that is bad.

Reply:
it's probably enough to just let the compiler inline it when appropriate (which means marking with `#[inline]` to support inlining across crates, which is really useful for these low-level, public methods).